### PR TITLE
Expression2 Effects

### PR DESCRIPTION
**Pull request by thegrb93**
Added effects. This is my first e2 customization so let me know if anything is wrong. I haven't encountered any bugs so far. The module also defaults to disabled so server owners can choose whether or not to have to deal with effect grief.

### CODE REVIEW

**Comment by AbigailBuccaneer** [comment]
:+1: Code looks sensible and clean to me.

It'd probably be worth a look over from somebody who actually uses E2 though :) @Divran maybe?

**Comment by thegrb93** [comment]
Crap, I meant to open a new pull request. How do I undo this.

**Comment by thegrb93** [comment]
I guess I can leave it as one pull request. I need some help with this commit. It works, but I can't find a way to update the egp once the loading is done so I have to create the egp again after all the textures are downloaded. Also I think I may need to add a way to clean/limit the textures that aren't used.

**Comment by thegrb93** [comment]
Added support to play url sounds and added the convar wire_expression2_sound_allowurl

**Comment by thegrb93** [comment]
Okay I found out how to do different pull requests. I'll do that instead of putting them all here.

File: lua/entities/gmod_wire_egp/lib/egplib/usefulfunctions.lua [hunk 1]
```diff
@@ -155,12 +155,69 @@ function EGP:SendMaterial( obj ) -- ALWAYS use this when sending material
...
+				["$vertexcolor"] = 1,
+				["$vertexalpha"] = 1,
+				["$ignorez"] = 1,
+				["$nolod"] = 1,
+				}
+				local urlmaterial = CreateMaterial("egp_urltex_" .. util.CRC(mat .. SysTime()), "UnlitGeneric", ShaderInfo)
+				local tex = v.Panel:GetHTMLMaterial():GetTexture("$basetexture")
+				tex:Download()
+				urlmaterial:SetTexture("$basetexture", tex)
+				tex:Download()
```
Comment (Nebual) [inline lua/entities/gmod_wire_egp/lib/egplib/usefulfunctions.lua]: Is this second tex:Download() intentional?

Comment (thegrb93) [inline lua/entities/gmod_wire_egp/lib/egplib/usefulfunctions.lua]: I don't know. I found that snippet in CapsAdmin's Pac3 mod. I was also confused and I can't find any documentation on it.

**Comment by Nebual** [comment]
Seems useful; its certainly a better designed approach than the 10 parameter behemoth function I've been using so far.

Example E2 code:
```
@name Water Effect Test
@persist Plays Water:effect

if(first()) {
    Water = effect()
    Water:setOrigin(entity():pos())
    Water:setNormal(vec(0,0,1))
    Water:setScale(350)
    Plays = 4
}
if(Plays > 0) {
    Plays--
    Water:play("watersplash")
    interval(250)
}
```
Just a thought, it might be handy to ```return this``` in each of the modifier functions to enable method chaining, which could shorten that init function to:
```
if(first()) {
    Water = effect():setOrigin(entity():pos()):setNormal(vec(0,0,1)):setScale(350)
    Plays = 4
}
```

Either way I approve.

**Comment by thegrb93** [comment]
Sure, I can do that real quick.

**Comment by Divran** [comment]
Just one concern. Wouldn't it be pretty easy to FPS rape everyone on the server with this? There are currently no limits to how fast you can spam effects. Maybe you should consider adding a replenishing buffer, like many other E2 extensions have. What it does is this
```
* On E2 initialize, set a number to its max amount.
* Every time you use the play function, subtract one from this number.
* If the number is zero, don't allow the user to use the play function.
* Every interval (pick a good interval, maybe 0.1 seconds or something), add 1 to the number, until it reaches its max amount again.
```
This allows you to create many effects at the same time, until the number is zero, or it allows you to create effects at a steady rate, as long as you do it at the same rate as the number is increased.

Here's an example of this: https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/debug.lua#L15 (lines 15 to 55) This implementation, used by most print functions in E2, also allows the users to set their own max limit clientside, but since effect creation is serverside, I don't think you could do that in this case, unless there's a way to create effects for specific players.

**Comment by Nebual** [comment]
Should also add a large ops cost to ef:play(), possibly 30 to bring it in line with propSpawn.

**Comment by Divran** [comment]
propSpawn should have a cost of at least 300. Can't believe it's so low.

**Comment by thegrb93** [comment]
I could make the limit for individual e2s, same way the sound plugin is.

**Comment by Divran** [comment]
Whether or not it's per-E2 or per player is a good point. I suppose it's up to you, but just remember that If you make it per-E2, then you can get twice as many effects by having twice as many E2s.

**Comment by AbigailBuccaneer** [comment]
Yeah, I think per-player makes more sense.

**Comment by thegrb93** [comment]
Okay I can do that. Should I change the sound system to be the same way?

**Pull request merged**